### ISSUE TITLE
False positive report of "drop this else and outdent its block"

### ISSUE THREAD

**Issue by rameshvk** [golang/lint#354]
Running golint against the following code causes the error `if block ends with a return statement, so drop this else and outdent its block` on L17.

This is spurious.  What is strange is that all the "if" clauses seem to be needed for this error to reproduce.  I tried removing any one of the if clauses and the error does not happen.

```go
func golintBug(x, y, z int) int {
        result := 0

        if x < 10 {
                result = 4
        } else if x < 20 {
                result = 3
        } else if x < 30 {
                return 2
        } else if x < 40 {
                return 1
        } else if x < 50 {
		return 0
        } else { // golint warning on this line
                result = 5
        }
        return result * x
}
```

**Comment by rameshvk** [comment]
Note: the actual code here is obviously not doing anything useful but it is a stripped down version of a real function in my codebase.  Happy to provide a link to the real code if needed.

**Comment by kevinburke** [comment]
Another snippet from x/crypto/openpgp/keys.go:

```go
		case *packet.Signature:
			if pkt.SigType == packet.SigTypeKeyRevocation {
				revocations = append(revocations, pkt)
			} else if pkt.SigType == packet.SigTypeDirectSignature {
				// TODO: RFC4880 5.2.1 permits signatures
				// directly on keys (eg. to bind additional
				// revocation keys).
			} else if current == nil {
				return nil, errors.StructuralError("signature packet found before user id packet")
			} else {
				current.Signatures = append(current.Signatures, pkt)
			}
```

yields:

```
openpgp/keys.go:385:11: if block ends with a return statement, so drop this else and outdent its block
```

**Comment by kevinburke** [comment]
Here's the code in question. I'm confused why the last block is an `*ast.ReturnStmt`.

```
// lintElses examines else blocks. It complains about any else block whose if block ends in a return.
func (f *file) lintElses() {
	// We don't want to flag if { } else if { } else { } constructions.
	// They will appear as an IfStmt whose Else field is also an IfStmt.
	// Record such a node so we ignore it when we visit it.
	ignore := make(map[*ast.IfStmt]bool)

	f.walk(func(node ast.Node) bool {
		ifStmt, ok := node.(*ast.IfStmt)
		if !ok || ifStmt.Else == nil {
			return true
		}
		if ignore[ifStmt] {
			return true
		}
		if elseif, ok := ifStmt.Else.(*ast.IfStmt); ok {
			ignore[elseif] = true
			return true
		}
		if _, ok := ifStmt.Else.(*ast.BlockStmt); !ok {
			// only care about elses without conditions
			return true
		}
		if len(ifStmt.Body.List) == 0 {
			return true
		}
		shortDecl := false // does the if statement have a ":=" initialization statement?
		if ifStmt.Init != nil {
			if as, ok := ifStmt.Init.(*ast.AssignStmt); ok && as.Tok == token.DEFINE {
				shortDecl = true
			}
		}
		lastStmt := ifStmt.Body.List[len(ifStmt.Body.List)-1]
		if _, ok := lastStmt.(*ast.ReturnStmt); ok {
			extra := ""
			if shortDecl {
				extra = " (move short variable declaration to its own line if necessary)"
			}
			f.errorf(ifStmt.Else, 1, link(styleGuideBase+"#indent-error-flow"), category("indent"), "if block ends with a return statement, so drop this else and outdent its block"+extra)
		}
		return true
	})
```

**Comment by kevinburke** [comment]
cc @dsymonds

**Comment by rameshvk** [comment]
@kevinburke thanks for that snippet.  I don't understand the code well enough but would hazard a guess that the problem is that the `ignore` flag is not getting propagated to all the else's in a chain because of the early return via `if ignore[ifStmt] { return true }`.

Just moving that if-clause to be after the next if-clause (i.e. swapping that and the next if-clause) causes my error to go away.

**Comment by rameshvk** [comment]
Also confirmed that swapping the two if-clauses like so:

```go
		if elseif, ok := ifStmt.Else.(*ast.IfStmt); ok {
			ignore[elseif] = true
			return true
		}
		if ignore[ifStmt] {
			return true
		}
```

fixes both false positives reported on this thread. 

I don't think the project is actively maintained though or I would go through the hassle of signing the CLA to get the fix in.

**Comment by kevinburke** [comment]
I've signed the CLA, do you mind if I submit it and give you credit?

**Comment by rameshvk** [comment]
@kevinburke go for it -- no credit needed!

**Comment by bcmills** [comment]
Consider this input (courtesy of @alandonovan):

```go
if … {
	return
} else if … {
	// don't return
} else if … {
	return
} else {
	…
}
```

Today, it exhibits the wrong-line warning as reported.

Now consider the equivalent structure expressed using a Block rather than a nested IfStmt:

```go
if … {
	return
} else {
	if … {
		// don't return
	} else {
		if … {
			return
		} else {
			…
		}
	}
}
```

Ideally, lint should produce a set of suggestions that lead to the same (warning-free) end state for the two equivalent lexical structures.  That would imply a warning along the lines of "else-block contains a single if-statement, so drop the block and outdent it", which a hypothetical user might follow with preference over the "drop this else-block" warning.

To my knowledge, no such warning currently exists.

----

That leads to a question: what is the purpose of this lint warning? Is it to flag unclear control-flow, or to flag over-indentation?

If the purpose is to flag over-indentation, we should also flag over-indentation for `default` blocks in switches from which every case returns, and we should flag loop control keywords (`break` and `continue`). To my knowledge we do not, and I don't see any issue reports for those.

So I'm left to conclude that lint is flagging unclear control-flow, not over-indentation per se, and in that regard it seems to be correct here: I really do find the examples described here needlessly difficult to follow as written.

----

I would argue that the example from https://github.com/golang/lint/issues/354#issuecomment-347603020 would be clearer as:
```go
if pkt.SigType == packet.SigTypeKeyRevocation {
	revocations = append(revocations, pkt)
} else if pkt.SigType == packet.SigTypeDirectSignature {
	// TODO: RFC4880 5.2.1 permits signatures
	// directly on keys (eg. to bind additional
	// revocation keys).
} else {
	if current == nil {
		return nil, errors.StructuralError("signature packet found before user id packet")
	}
	current.Signatures = append(current.Signatures, pkt)
}
```

or, better still:
```go
switch pkg.SigType {
case packet.SigTypeKeyRevocation:
	revocations = append(revocations, pkt)
case packet.SigTypeDirectSignature:
	// TODO: RFC4880 5.2.1 permits signatures
	// directly on keys (eg. to bind additional
	// revocation keys).
default:
	if current == nil {
		return nil, errors.StructuralError("signature packet found before user id packet")
	}
	current.Signatures = append(current.Signatures, pkt)
}
```  

Applying that same transformation to the code in https://github.com/golang/lint/issues/354#issue-273565116, we find that the problem is lint *under-warning*, not *over-warning*.

```go
func golintBug(x, y, z int) int {
        result := 0

        if x < 10 {
                result = 4
        } else if x < 20 {
                result = 3
        } else if x < 30 {
                return 2
        } else if x < 40 {  // LINT: if block ends with a return statement, so drop this else and outdent its block
                return 1
        } else if x < 50 {  // LINT: if block ends with a return statement, so drop this else and outdent its block
		return 0
        } else {  // LINT: if block ends with a return statement, so drop this else and outdent its block
                result = 5
        }
        return result * x
}
```

With warnings correctly applied at all of the relevant points, a clearer (and correct) fix becomes apparent:
```go
func golintBug(x, y, z int) int {
        result := 0

        if x < 10 {
                result = 4
        } else if x < 20 {
                result = 3
        } else {
                if x < 30 {
                        return 2
                }
                if x < 40 {
                        return 1
                }
                if x < 50 {
                        return 0
                }
                result = 5
        }
        return result * x
}
```

or perhaps:
```go
func golintBug(x, y, z int) int {
        result := 0

        switch {
        case x < 10:
                result = 4
        case x < 20:
                result = 3
        case x < 30:
                return 2
        case x < 40:
                return 1
        case x < 50:
                return 0
        default:
                result = 5
        }
        return result * x
}
```

**Comment by rameshvk** [comment]
I think that is taking `opinionated` a bit far :)

FWIW, I did switch my real world use case to use `switch{}` but there are other cases where the warning didn't make sense at all.  I do workaround them but it leaves me with a bad feeling about golint.

**Comment by alandonovan** [comment]
> Ideally, lint should produce a set of suggestions that lead to the same (warning-free) end state for the two equivalent lexical structures.

There are many high-level control structures that can express a given control-flow graph and set of variable declarations, and I'm not convinced an algorithm can reliably pick the clearest. Sometimes it's desirable to declare a block just to hold a comment that documents a no-op, for example.

**Comment by kevinburke** [comment]
I believe the intent was to flag over-indentation. The code as written does not handle too many cases, and specifically tries to avoid issuing a warning in situations with multiple if-else's, it's just currently doing it incorrectly.

**Comment by dsymonds** [comment]
I wrote the original check, so I know what the intent was, and it was indeed to discourage over-indentation. Go favours a style like

```go
if something {
  return
}
if something_else {
  return
}
do_something()
```

instead of what might be more common elsewhere as

```go
if !something {
  if !something_else {
    do_something()
  }
}
```

In the case of the if-else check, it's to discourage constructs like

```go
if foo {
  return x
} else {
  return y
}
```

since most of the time one of those branches is less common or handling an error. The Go style is to instead write that as

```
if foo {
  return x
}
return y
```

so that the most common or "normal" control flow stays outdented. This is most common when there's a trailing else (with no condition), though it clearly generalises a bit from that.

From the original report, this is merely a bug in golint's implementation. It should only be suggesting dropping the else when all the previous if branches end with a return statement.

I agree with most of the remainder of what @bcmills wrote, though. However, the simple chain of if/else-if/else was all that was originally intended to cover, rather than trying to be more comprehensive about flagging suboptimal control flows.

**Comment by gelldur** [comment]
Other false positive:
```
func (g *Gaga) getFunnyStuff(source, target string) (Attributes, error) {
	if attributes, err := g.GetSong(); err == nil {
		return *attributes, err
	} else {
		return nil, err
	}
}
```

**Comment by bcmills** [comment]
@gelldur, that example is not a false positive. It is exactly the sort of code that the check is intended to flag.

The if-statement would be idiomatically written:
```go
	attributes, err := g.GetSong()
	if err != nil {
		return nil, err
	}
	return *attributes, err
```

**Comment by rameshvk** [comment]
The original issue seems to be fixed.  Closing this issue.

**Comment by TomasB** [comment]
> I wrote the original check, so I know what the intent was

Did you account though that the warning will make  'If with a short statement' useless?

We can't use this now, which is very common pattern: 
```
if v, err := f(); err {
   log.Error(err)
} else {
   mem = v
}
```

Instead, we have to write:
```
v, err := f();
if err {
   log.Error(err)
}
mem = v
```

which isn't as clean, since we expanded scope for v, which might lead to errors.

Given the choice, we (devs) would rather take if with short statement over that warning.
Number of opened tickets for that erroneous warning supports my statement.

**Comment by gunjanmimo** [comment]
if we have multiple lines in else part how we ca element else ?
`	if err != nil {
		if os.IsNotExist(err) {
			logrus.Error(err)
			return "", "", err
		} else {
			err = errors.Wrap(err, "unknown error")
			logrus.Error(err)
			return "", "", err
		}
	}`

**Comment by bcmills** [comment]
@gunjanmimo 
```go
	if err != nil {
		if !os.IsNotExist(err) {
			err = errors.Wrap(err, "unknown error")
		}
		logrus.Error(err)
		return "", "", err
	}
```
